### ISSUE TITLE
Move dependencies into package.json dependencies

### ISSUE THREAD

**Issue by matthewp** [bit-docs/bit-docs#56]
The fact that bit-docs installs its own dependencies makes it more difficult to debug plugins. You cannot go into node_modules and modify a plugin and rerun the build. It would be better if the package manager did the installing and bit-docs did not.

**Comment by matthewp** [comment]
This is especially difficult because even if you go into `node_modules/bit-docs/lib/configure/node_modules` and modify code, you cannot rebuild while using that code. You get a freshly downloaded copy each time.

**Comment by matthewp** [comment]
Looking at other tools, babel works with plugins without installing them itself. Instead you list the plugins in your dependencies and reference just the package names in the babel config. I think it makes sense for bit-docs to work that way as well.

**Comment by justinbmeyer** [comment]
This was done because bit-docs plugins were conflicting with normal `package.json` dependencies.  Also, bit-docs deps can be "big" and I wanted to avoid requiring installing them in every development workflow.

To debug, there's `devBuild: true` option.  Also, it's possible to setup a plugin test harness like `bit-docs-html-canjs`, where it installs itself.  I can share this example more if interested.

**Comment by justinbmeyer** [comment]
I do strongly think there needs to be a way to rebuild without re-installing.  It's not hard to add, I just haven't had time ... maybe tonight ...

**Comment by matthewp** [comment]
Can you describe the problem exactly? I don't quite follow. Many other building type of tools with plugins work by installing plugin as normal npm dependencies. Some examples: babel, eslint.

**Comment by justinbmeyer** [comment]
I don't remember specifically why this was done, I do remember it was intentional / purposeful after trying to make plugins normal npm dependencies.

I believe it had to do with plugins and normal dependencies conflicting.  CanJS's (or maybe some app's) `package.json` was something like:

```
bit-docs-widget-using-can-control: "^0.0.1"
can-control: ">2"
```

if `bit-docs-widget-using-can-control`'s `package.json` had something like:

```
can-control: "2.1.1"
```

Then npm might have been installing `2.1.1` for CanJS instead of the latest `can-control`.  I don't think this was the exact situation, but it was something like it.

Similarly, `bit-docs` uses itself to document itself. I might want to use `bit-docs-plugin@2.1.1` to document `bit-docs-plugin@2.1.3`.

**Comment by justinbmeyer** [comment]
It might also have to do with being able to do a `-f` build without blowing away everything in `node_modules`.  I think if we shared a `node_modules` folder, it might be hard to do a total refresh without re-installing everything again ...

**Comment by matthewp** [comment]
Why would `-f` blow away node_modules in this scenario? This proposal is that bit-docs not do installs at all.

It sounds like there isn't a clear reason for this implementation. It's possibly that constraints of how npm installed at the time were part of the equation as well. I think we should try again and see what happens. As I said upthread, other tools are able to get by without install their own plugins and I see no difference that would require bit-docs to do so.

**Comment by justinbmeyer** [comment]
> Similarly, bit-docs uses itself to document itself. I might want to use bit-docs-plugin@2.1.1 to document bit-docs-plugin@2.1.3.

I think that's an important reason.  Is babel built using itself?  How?

**Comment by matthewp** [comment]
Yes, it is, but my guess is that it doesn't install an old version of itself to do so, it just uses its own, current, code. I'm not sure why you would ever use an old version of yourself (a package). Even still, it seems like it should work nonetheless. When you run `require("bit-docs-plugin")` it will look in your package.json to resolve that, you cannot require yourself by name, in this way.

**Comment by justinbmeyer** [comment]
Another reason I think  ... is that there are "node" and "client" dependencies.  Plugins can label themselves as needing to be installed into `bit-docs-generate-html`, which builds a separate package.json. Investigating this a bit more...

**Comment by matthewp** [comment]
I ran into this problem again trying to fix this bug: https://github.com/bit-docs/bit-docs-tag-package/issues/6

This is a huge issue and not something any other plugin based system suffers from, to my knowledge.

**Comment by matthewp** [comment]
Another problem this creates is that `npm ls` doesn't work on these installed packages, that information is completely opaque and it makes it harder to trace the dependency graph of plugins.
